### ISSUE TITLE
ImGui dockable window with its own native window/context

### ISSUE THREAD

**Issue by mnesarco** [ocornut/imgui#4613]
_Config/Build Information_
```
Dear ImGui 1.85 WIP (18412)
--------------------------------
sizeof(size_t): 8, sizeof(ImDrawIdx): 2, sizeof(ImDrawVert): 20
define: __cplusplus=202002
define: IMGUI_DISABLE_OBSOLETE_FUNCTIONS
define: __linux__
define: __GNUC__=4
define: __clang_version__=12.0.1 
define: IMGUI_HAS_VIEWPORT
define: IMGUI_HAS_DOCK
```

Version: 1.85 WIP (18412)
Branch: docking

```
io.BackendPlatformName: imgui_impl_glfw
io.BackendRendererName: imgui_impl_opengl3
io.ConfigFlags: 0x00000000
io.ConfigViewportsNoDecoration
io.ConfigInputTextCursorBlink
io.ConfigWindowsResizeFromEdges
io.ConfigMemoryCompactTimer = 60.0
io.BackendFlags: 0x0000140E
 HasMouseCursors
 HasSetMousePos
 PlatformHasViewports
 RendererHasVtxOffset
 RendererHasViewports
```

**My Issue/Question:**

I want to use imgui with another library that renders to the entire native window, so currently I am rendering imgui on top of the of the previously rendered content, but I want to confine the other lib rendering into one dockable ImGui window. I know this can be done with framebuffers or something like that but the other library is quite complicated so the easy path is to keep it rendering to the whole window. I know that the multi-viewport branch creates windows with its own native window when they are dragged outside of the main viewport, but I don't want to drag anything outside, i want to have a dockable window with its native window managed by ImGui so I can pass the native window handle to the library to render. Is it possible?

**Screenshots/Video**

![screenshot_24](https://user-images.githubusercontent.com/1087317/135720661-c090eee6-b493-4d60-abcd-7a591f337080.png)

**Standalone, minimal, complete and verifiable example:** _(see https://github.com/ocornut/imgui/issues/2261)_
```
// That's the question :)
```

**Comment by PathogenDavid** [comment]
I feel like any solution to this problem that doesn't involve properly convincing your other library to render to a frame buffer is just going to involve rendering to a separate framebuffer with more steps.

**Comment by mnesarco** [comment]
@PathogenDavid you are right. I need to fight with the lib and go with the FrameBuffer

**Comment by meshula** [comment]
Dear Imgui has a mechanism for this. See this section in imgui.h.

```cpp
// ImDrawCallback: Draw callbacks for advanced uses [configurable type: override in imconfig.h]
// NB: You most likely do NOT need to use draw callbacks just to create your own widget or customized UI rendering,
// you can poke into the draw list for that! Draw callback may be useful for example to:
//  A) Change your GPU render state,
//  B) render a complex 3D scene inside a UI element without an intermediate texture/render target, etc.
// The expected behavior from your rendering function is 'if (cmd.UserCallback != NULL) { cmd.UserCallback(parent_list, cmd); } else { RenderTriangles() }'
// If you want to override the signature of ImDrawCallback, you can simply use e.g. '#define ImDrawCallback MyDrawCallback' (in imconfig.h) + update rendering backend accordingly.
#ifndef ImDrawCallback
typedef void (*ImDrawCallback)(const ImDrawList* parent_list, const ImDrawCmd* cmd);
#endif

// Special Draw callback value to request renderer backend to reset the graphics/render state.
// The renderer backend needs to handle this special value, otherwise it will crash trying to call a function at this address.
// This is useful for example if you submitted callbacks which you know have altered the render state and you want it to be restored.
// It is not done by default because they are many perfectly useful way of altering render state for imgui contents (e.g. changing shader/blending settings before an Image call).
#define ImDrawCallback_ResetRenderState     (ImDrawCallback)(-1)
```

**Comment by mnesarco** [comment]
Thank you @meshula , but i think ImDrawCallback is not related with my problem. I have a library that received the gl context and do render in an opaque function using all the native window space.

Anyway I have managed to implement the FrameBuffer and faked the window object so the lib "think" it has the window. The only think I don't like is that I have to recreate the FBO everytime the ImGui window resizes, but I think this is not avoidable.

**Comment by mnesarco** [comment]
As I have to recreate the FrameBuffer on every ImGui Window resize, it would be great to do it only after the resize have finished. Currently if the user resizes the window dragging the handle, the window size changes a lot of times per second. I couldn't find any way to detect the "resizing" state. Does someone know a way to detect the resizing state of an ImGui Window?

**Comment by mnesarco** [comment]
As suggested in another Issue, I am creating the FrameBuffer with the size of the entire OS window and rendering only the region required by the ImGui window. It appears as a waste of memory but actually the ImGui Window is docked at the center node and covers a big part of the OS Window and this way I only recreate the FBO when the OS window resizes (Very rare). I think this solution is ok. But any further suggestion/comment are welcome.
